### PR TITLE
hikey*: support migration to BL2_EL3 on arm-tf

### PR DESCRIPTION
**Pull request by vchong**
Signed-off-by: Victor Chong <victor.chong@linaro.org>

### CODE REVIEW

**Comment by vchong** [comment]
Ok now. Please review.

**Comment by vchong** [comment]
In order to test, please `repo sync`, `make cleaner` and rebuild.

Thanks!

**Comment by vchong** [comment]
@fboudra fyi only

**Comment by jbech-linaro** [comment]
I've built HiKey (old one) using this patch and build is fine. However, since I have no HiKey to spare I cannot try it out on real hardware.

For HAB, I checked what is wrong:
```bash
Expecting: ['HEAD is now at ba11e24', <class 'pexpect.exceptions.TIMEOUT'>] (timeout=20, error=1)
Got: 1 (error at 1)      
Returning STATUS_FAIL    
Failed to checkout ba11e24b9f18232ff2139cddd0cd4ca4122f8752                                          
Failed building job 
```
For some reason it couldn't find the commit on Vee's branch (hilarious start of the commit, btw). When Vee squash and force push a new patch it will try build it again. Let's hope that works better.

`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by vchong** [comment]
@jbech-linaro I was trying to be efficient (or lazy depending on how you `ba11e24b9f18232ff2139cddd0cd4ca4122f8752` was an overwritten commit in my branch. I was forcing pushes pretty quickly before finalizing the PR. Maybe the HAB hook and build job wasn't able to 'catch up'?

**Comment by vchong** [comment]
@jbech-linaro The flash and test log looked ok though. I was mainly forcing pushes for hikey960.

**Comment by vchong** [comment]
Fixup: 'fastboot flash loader' needs to be in recovery mode.

File: hikey.mk [hunk 1]
```diff
@@ -415,10 +416,11 @@ recovery:
 	@echo
 	$(call flash_help)
 	@echo
-	python $(ROOT)/burn-boot/hisi-idt.py --img1=$(LLOADER_PATH)/l-loader.bin
+	python $(ROOT)/burn-boot/hisi-idt.py --img1=$(LLOADER_PATH)/recovery.bin
+	fastboot flash loader $(LLOADER_PATH)/l-loader.bin
 	@echo
 	@echo "3. Wait until you see the (UART) message"
-	@echo "    \"Enter downloading mode. Please run fastboot command on Host.\""
+	@echo "    \"Enter fastboot mode. Please run fastboot command on Host.\""
```
Comment (jforissier) [inline hikey.mk]: It's not the message I'm seeing. After booting with the jumpers in recovery mode, and running the Python script, the console is like this:
```
NOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v1.4(release):v1.5-rc2
NOTICE:  BL1: Built : 10:05:55, Mar  9 2018
NOTICE:  BL1-FWU: *******FWU Process Started*******
NOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v1.1(release):af5ddb1
NOTICE:  BL1: Built : 10:04:04, Mar  9 2018
NOTICE:  syspll frequency:1190494208Hz
NOTICE:  succeed to init lpddr3 rank0 dram phy
INFO:    lpddr3_freq_init, set ddrc 533mhz
INFO:    init ddr3 rank0
INFO:    ddr3 rank1 init pass
INFO:    lpddr3_freq_init, set ddrc 800mhz
INFO:    init ddr3 rank0
INFO:    ddr3 rank1 init pass
INFO:    Elpida DDR
NOTICE:  Enter fastboot mode...
```

And AFAICT there is no need to wait. For me at least I can run the fastboot commands immediately, they'll just wait for the board to be ready if it's not.

**Comment by jbech-linaro** [comment]
@vchong it shouldn't be an issue, it should end up in the job queue, but there are for sure both bugs and room for improvements. So you probably found a sequence making HAB go foobar. Anyhow, I manually re-delivered the final payload to trigger the last push again and now HAB seems happy with it all. Good job!

File: hikey_debian.mk [hunk 2]
```diff
@@ -463,7 +463,7 @@ recovery:
 	python $(ROOT)/burn-boot/hisi-idt.py --img1=$(LLOADER_PATH)/l-loader.bin
```
Comment (jforissier) [inline hikey_debian.mk]: I suppose `hikey_debian.mk` needs the same changes as `hikey.mk`.

**Comment by jforissier** [comment]
Tested "make recovery" and "make flash" on both HiKey and HiKey960 (but not HiKey Debian).
Caveat: I have tested these patches in my [optee_build](https://github.com/jforissier/optee_build) environment, not with build.git/manifest.git.

`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960)`

File: hikey.mk (hunk 1 above)
Comment (vchong) [reply hikey.mk]: Updated UART message. Thanks.

File: hikey_debian.mk (hunk 2 above)
Comment (vchong) [reply hikey_debian.mk]: Oops yes. Will fix. Thanks.

File: hikey960.mk [hunk 3]
```diff
@@ -409,11 +411,12 @@ endif
...
 	@echo "     Press RETURN or SPACE key to quit.\""
+	@echo "It can take a while for the fastboot device to come up,"
+	@echo "so please wait ~10 seconds."
 	@read -r -p "Then press enter to continue flashing" dummy
 	@echo
 	fastboot flash ptable $(LLOADER_PATH)/prm_ptable.img
 	fastboot flash xloader $(IMAGE_TOOLS_PATH)/sec_xloader.img
 	fastboot flash fastboot $(LLOADER_PATH)/l-loader.bin
 	fastboot flash fip $(ARM_TF_PATH)/build/hikey960/$(ARM_TF_BUILD)/fip.bin
-	fastboot flash nvme $(IMAGE_TOOLS_PATH)/nvme.img
```
[the remark this answers is not in the document]
Comment (vchong) [reply hikey960.mk]: It's legacy and not needed for proper operation anymore but looks like it'll still generate some warning text if removed. Will add it back. Note that the file has been renamed tho `hisi-nvme.img` though. Thanks.

**Comment by vchong** [comment]
@jforissier All comments addressed. Please continue review. Thanks.

**Comment by jforissier** [comment]
Thanks Victor.
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by vchong** [comment]
Nope, thank you!
Rebased, squashed and tags added.

**Pull request merged**